onnxscript/version_converter/version_converter.py

```diff
@@ -1,55 +1,427 @@
+        self.op_adapters: dict[tuple[str, str], list[AdapterVersionChecker]] = {}
+
+    def lookup_adapters(
+        self, domain: str, opname: str, original_version: int, target_version: int
+    ):
+        adapter_list = self.op_adapters.get((domain, opname), [])
+        return [
+            adapter.function
+            for adapter in adapter_list
+            if adapter.valid_for(opname, original_version, target_version)
+        ]
+
+    def register(
+        self, opname: str, domain: str = "", node_version=None, upgrade_version=None
+    ) -> Callable[[AdapterFunction], AdapterFunction]:
+        if (domain, opname) in self.op_adapters:
+            adapter_list = self.op_adapters[(domain, opname)]
+        else:
+            adapter_list = []
+            self.op_adapters[(domain, opname)] = adapter_list
+        if node_version is None or upgrade_version is None:
+            original_version = None
+            target_version = None
+        else:
+            original_version = node_version
+            target_version = upgrade_version
+
+        def decorator(function: AdapterFunction) -> AdapterFunction:
+            adapter_list.append(
+                AdapterVersionChecker(original_version, target_version, function)
+            )
+            return function
+
+        return decorator
+
+
+registry: AdapterRegistry = AdapterRegistry()
+
+register = registry.register
+
+
+def _get_input(node: ir.Node, index: int) -> ir.Value | None:
+    if index < len(node.inputs):
+        return node.inputs[index]
+    return None
+
+
+def _get_output(node: ir.Node, index: int) -> ir.Value | None:
+    if index < len(node.outputs):
+        return node.outputs[index]
+    return None
+
+
+def _get_int_attribute(node: ir.Node, name: str, default: int | None = None) -> int | None:
+    if name in node.attributes:
+        attr = node.attributes[name]
+        if not isinstance(attr, ir.Attr):
+            return None
+        attr_val = attr.value
+        if isinstance(attr_val, int):
+            return attr_val
+        # This is an invalid model: attribute has invalid/unexpected type.
+        # For now, we just return None. We could raise an error too.
+        return None
+    return default
+
+
+def _get_str_attribute(node: ir.Node, name: str, default: str | None = None) -> str | None:
+    if name in node.attributes:
+        attr = node.attributes[name]
+        if not isinstance(attr, ir.Attr):
+            return None
+        attr_val = attr.value
+        if isinstance(attr_val, str):
+            return attr_val
+        # This is an invalid model: attribute has invalid/unexpected type.
+        # For now, we just return None. We could raise an error too.
+        return None
+    return default
+
+
+## Op-specific adapters
+
+# Opset 19 -> 20
+
+
+@register("DFT", node_version=19, upgrade_version=20)
+def dft_19_20(node: ir.Node, op):
+    input = node.inputs[0]
+    inverse = _get_int_attribute(node, "inverse", 0)
```

Comment:
There is a complication here, relating to handling of functions. In functions, we can have attribute-parameters, which mean we don't know the value of the attribute. The version-conversion, as written, will be wrong in such cases.

For now, I think it is okay to inline all functions (see relevant comment below) and ignore this issue.

Reply:
Would it be possible to just promote the attribute to inputs and use them in the upgraded op?